model_compression_toolkit/keras/gradient_ptq/gptq_training.py

```diff
@@ -0,0 +1,144 @@
+
+        if float_user_info.input_scale != gptq_user_info.input_scale:
+            common.Logger.error("Input scale mismatch between float and GPTQ networks")  # pragma: no cover
+        else:
+            self.input_scale = gptq_user_info.input_scale
+
+    def train(self, representative_data_gen: Callable):
+        """
+        Train the quantized model using GPTQ training process in Keras framework
+        Args:
+            representative_data_gen: Dataset to use for inputs of the models.
+        """
+
+        def update_step(input_data: List[np.ndarray]) -> Tuple[tf.Tensor, List[tf.Tensor]]:
+            """
+            Get outputs from both teacher and student networks. Compute the observed error,
+            and use it to compute the gradients and applying them to the student weights.
+            Args:
+                input_data: A list of Input tensors to pass through the networks.
+            Returns:
+                Loss and gradients.
+            """
+            y_float = self.float_model(input_data)  # running float model
+            with tf.GradientTape(persistent=True) as tape:
+                y_fxp = self.fxp_model(input_data)  # running fxp model
+                loss_value = self.gptq_config.loss(y_fxp, y_float, self.fxp_weights_list, self.flp_weights_list,
+                                              self.compare_points_mean, self.compare_points_std)
+
+            # Use the gradient tape to automatically retrieve
+            # the gradients of the trainable variables with respect to the loss.
+            grads = tape.gradient(loss_value, self.flattened_trainable_weights)
+
+            # Run one step of gradient descent by updating
+            # the value of the variables to minimize the loss.
+            self.gptq_config.optimizer.apply_gradients(zip(grads, self.flattened_trainable_weights))
+
+            return loss_value, grads
+
+        # ----------------------------------------------
+        # Training loop
+        # ----------------------------------------------
+        self.loss_list = []
+        for _ in tqdm(range(self.gptq_config.n_iter)):
+            data = representative_data_gen()
+            loss_value_step, grads = update_step([d * self.input_scale for d in data])
+            if self.gptq_config.log_function is not None:
+                self.gptq_config.log_function(loss_value_step, grads, self.flattened_trainable_weights, self.compare_points)
+            self.loss_list.append(loss_value_step.numpy())
+            common.Logger.debug(f'last loss value: {self.loss_list[-1]}')
+
+    def update_graph(self):
+        """
+        Update a graph using GPTQ after minimizing the loss between the float model's output
+        and the quantized model's outputs.
+        Returns:
+            Updated graph after GPTQ.
+        """
+        return update_graph_after_gptq(self.fxp_model,
```

Comment:
there is no reason to have the function: update_graph_after_gptq 
please move is content to this class

Reply:
Done